The first part of this notebook is dealing with the second milestone of the project. If you want to jump straight to Milestone 3, click [here](#m3).

# How to make or break a country

## Types of border changes

Before the start of our analysis we had to spend a significant amount of time exploring the border changes that happened in the period from 1979 until today, because these changes can be recognized from the data included in the *GDELT* dataset.
After detailed investigation of the possible reasons for the changes of borders, we found out that there are 5 situations that resulted in the alterations of the territorial integrity:

 * the decomposition of large socialist/communist countries – Most countries in the period of the last (almost) 40 years were created after the dissolution of two large socialist countries – Soviet Union and Yugoslavia.
 * the last steps of decolonization – Although most of the colonies gained internationally recognized sovereignty in the end of the 19th century and the beginning of 20th century, some smaller countries in the Carribean got separated from the UK in the 1980s and some islands in Oceania gained their right for free association in the 1990s.
 * the separation after political or military conflict – Some countries faced the need for border changes as a result of political or military changes (for example Namibia). Some countries underwent the invasion of foreign forces that both started and ended in the time period we are observing, so they can also be included in this group as they could give different perspective to the problem we are analyzing. The example of occupation of Kuwait carried out by Iraqi forces in the 1990 is one of the representatives of this type.
 * the breakaway states – Contrary to the aforementioned types, some countries/areas have declared their independence, but the international public is not recognizing their proclamations as valid, even though some of them have functioned as the de facto independent or autonomous parts of other countries for over 30 years. One of the examples of this type is the Northern Cyprus area (officially the Turkish Republic of Northern Cyprus) which has proclaimed its independence in the 1983, but has since only been recognized by Turkey.
 * the unifications – Since our research will be focused toward all of the border changes that happened in the past 40 years, we decided that, although this type is a much rarer occurrence than the others, we should analyze the examples of the unification of  e.g. East and West Germany and North and South Yemen in the 1990s.
 
The main idea in the rest of the analysis will be to process all of the border changes together in order to check if all of them show some similar events that precede the changes, but then again in the later phases have some very different aspects according to the type of their change. In the end of the analysis, the question of currently on-going cases like Catalonia shall be analyzed to see if they show any compliance with the observed types.


## The observed countries

In order to decide which countries to use in our analysis we had to extract all of the countries that gained their sovereignty in the period that is covered in the GDELT dataset.

Firstly, we tried to create a list from all of the countries that became a part of the United Nations in the period of 1979-2017 from the [list of the member states of the United Nations](https://en.wikipedia.org/wiki/Member_states_of_the_United_Nations). Unfortunately, we couldn’t just extract the countries by date because there are some countries that existed for more than a century before officially becoming a member of the United Nations (e.g. Switzerland :)).

Instead we analyzed the dates presented in the tables of the [list of sovereign states by dates of formation](https://en.wikipedia.org/wiki/List_of_sovereign_states_by_date_of_formation), and kept the countries “created” in the required period. For each of the newly created countries we also collected the data about the previous states they were a part of, or from which they were unified. This part was mostly done by hand because we didn’t find any dataset that contains the information solely about the countries we are focusing on and usually they don’t contain any data about the predecessor countries. Also, the dates in the *Wikipedia* tables we used are spread across several columns and need human interpretation in order to be correctly extracted.

Even though we listed the new countries we will examine, our problems were not solved yet. There was still the fact that the extracted list does not include any of the breakaway countries, since they are mostly not internationally recognized nor sovereign. We had to obtain these regions by analyzing the [list of breakaway states](https://en.wikipedia.org/wiki/Constituent_state#Breakaway_states) and than add to our list those that were created after 1979.

When we created the list of all of the countries/regions that we will use in our research, we had to find more details about their history and origins in order to make sure which type or types of border changes have they faced in the observed period.

## The types of border changes in the observed countries

The biggest change of borders caused by a single event in our analysis will definitely be the dissolution of the [Soviet Union](https://en.wikipedia.org/wiki/Soviet_Union). In the period from March 1990 to the December of 1991, 15 new countries were created from the former Soviet Union and these are:
 * Russia – the official successor state,
 * Armenia,
 * Azerbaijan,
 * Georgia,
 * Kazakhstan,
 * Kyrgyzstan,
 * Tajikistan,
 * Turkmenistan,
 * Uzbekistan,
 * Belarus,
 * Estonia,
 * Latvia,
 * Lithuania,
 * Moldova,
 * Ukraine.
 
All of these countries can be classified as the results of the first type of border changes we mentioned – the decomposition of large socialist/communist countries.

The listed 15 countries do not mark not the end of the decomposition, because in the following years 5 more breakaway states proclaimed their independence (with limited international recognition) on the territories of the former Soviet Union:
 * [Artsakh](https://en.wikipedia.org/wiki/Republic_of_Artsakh) (previously known as Nagorno-Karabakh) – area internationally recognized as a part of Azerbaijan, declared independence in 1991 and functions as the de facto part of Armenia,
 * [Abkhazia](https://en.wikipedia.org/wiki/Abkhazia) – area internationally recognized as a part of Georgia, gained de facto independence in 1993, but is considered to be a Russian-occupied territory by the Georgian government,
 * [South Ossetia](https://en.wikipedia.org/wiki/South_Ossetia) – area that declared independence from Georgia in 1991, to this day is under de facto control of Ossetian and Russian forces, so a significant part of the international community considers it to be occupied by Russia, but is not recognized as a country nor administrative area by the Georgian government,
 * [Transnistria](https://en.wikipedia.org/wiki/Transnistria) – area internationally considered to be a part of Moldova, gained de facto independence in 1990, but is mostly populated by Russian people,
 * [Chechen Republic of Ichkeria](https://en.wikipedia.org/wiki/Chechen_Republic_of_Ichkeria) (also known as [Chechnya](https://en.wikipedia.org/wiki/Chechnya)) – gained de facto independence in 1996, but was returned under the rule of Russia in 2000.

These regions are considered to be the breakaway states.

Many countries of interest for our research originated from other large socialist country – [Yugoslavia](https://en.wikipedia.org/wiki/Yugoslavia). There are 6 states that were created after the dissolution of Yugoslavia that started in 1991:
 * Serbia – the official successor state (part of [Serbia and Montenegro](https://en.wikipedia.org/wiki/Serbia_and_Montenegro) until 2006),
 * Montenegro (part of [Serbia and Montenegro](https://en.wikipedia.org/wiki/Serbia_and_Montenegro) until 2006),
 * Croatia,
 * Bosnia and Herzegovina,
 * Slovenia,
 * Macedonia (formally associated as the former Yugoslav Republic of Macedonia).

All of the mentioned countries can again be associated with the border change type mentioned as the decomposition of large socialist/communist countries.

Additionally, in 2008 [Kosovo](https://en.wikipedia.org/wiki/Kosovo) unilaterally declared its independence from Serbia and gained partial diplomatic recognition as a sovereign state. This change is not recognized by Serbian government and this region will be considered a breakaway state in our analysis.

The last European socialist country that had a change of borders was [Czechoslovakia](https://en.wikipedia.org/wiki/Czechoslovakia) which got separated into Czech Republic and Slovakia in 1993. Because of that fact, both of these countries shall be observed like they belong to the type mentioned as the decomposition of large socialist/communist countries.

There is a number of Caribbean countries that gained their independence from the United Kingdom in the period from 1979 to 1983. Although these countries did remain a part of the [Commonwealth of Nations](https://en.wikipedia.org/wiki/Commonwealth_of_Nations) this period did mark the beginning of their internationally recognized sovereignty so it should be considered as a change of borders. All of these countries will be considered as they belong to the last steps of decolonization:
 * Antigua and Barbuda,
 * Belize,
 * Saint Kitts and Nevis,
 * Saint Lucia,
 * Saint Vincent and the Grenadine.

In the same group we can add the countries in Oceania Kiribati (gained independence from the United Kingdom in 1979) and Vanuatu (which gained independence in 1980 from United Kingdom and France). Brunei, a country in Southeast Asia, should also be classified in the last steps of decolonization group since it regained independence from United Kingdom in 1984. There is also Zimbabwe – an African country that declared unilateral independence from United Kingdom in 1965 under the name Rhodesia, but was officially granted sovereignty as Zimbabwe in 1980, which will be a part of our research in the same group.

The last countries that belong in the last steps of decolonization type of border changes are Marshall Islands, Federated States of Micronesia and Palau which became independent from USA by the [Compact of Free Association]( https://en.wikipedia.org/wiki/Compact_of_Free_Association) during the period from 1986 to 1994.

In the group that contains the countries which separated after political or military conflict, we added three countries from Africa: Eritrea (gained independence from Ethiopia in 1993), Namibia (gained independence from South Africa in 1990) and South Sudan (gained the independence most recently, from Sudan in 2011).

The Asian country Cambodia was occupied by Vietnamese troops from 1978 until 1991 when it regained independence as a result of the [Paris Comprehensive Peace Settlement](https://en.wikipedia.org/wiki/1991_Paris_Peace_Accords). Similarly, in 1990 Iraqi forces invaded and annexed Kuwait, but it regained sovereignty in 1991 after a military intervention. Both of these cases will be classified as the type of countries whose borders changed after political or military conflict.

Timor-Leste (known also as [East Timor](https://en.wikipedia.org/wiki/East_Timor)) first gained its very short-lived independence in 1975, but was invaded by Indonesia only 9 days later. This occupation lasted until 2002 when it became the first sovereign state od 21st century. Since that change of borders marked the end of an occupation, this case is also considered to be in the group of separations after political or military conflict.

There are also a few more breakaway states that we discovered during our research and which should be included in the later analysis. First one is Northern Cyprus ([Turkish Republic of Northern Cyprus](https://en.wikipedia.org/wiki/Northern_Cyprus)) which made the unilateral declaration of independence from Cyprus in 1983 but is still almost completely internationally unrecognized. There is also Palestine ([State of Palestine](https://en.wikipedia.org/wiki/State_of_Palestine)) which first declared independence in 1988 and has since gained partial international recognition, but still has a sovereignty dispute with Israel. We also investigated the example of Somaliland ([Republic of Somaliland](https://en.wikipedia.org/wiki/Somaliland)), a region internationally recognized as a part of Somalia which declared unilateral independence in 1991.

We discovered that Senegambia([Senegambia Confederation](https://en.wikipedia.org/wiki/Senegambia_Confederation)) was created by unification of Gambia and Senegal in 1982 in the West Africa. But this confederation only lasted until 1989 when two countries separated again. This example will be analyzed for both unification and separation after political or military conflict, for the periods that correspond to the different territorial organization the country had.

The most significant unification in the observed period is by far the reunification of East Germany ([German Democratic Republic](https://en.wikipedia.org/wiki/East_Germany)) and West Germany ([Federal Republic of Germany](https://en.wikipedia.org/wiki/West_Germany)) in 1990. The other example of unification is Yemen, which united [North Yemen](https://en.wikipedia.org/wiki/North_Yemen) and [South Yemen](https://en.wikipedia.org/wiki/South_Yemen) in 1990.

The case of Crimea (or [Republic of Crimea]( https://en.wikipedia.org/wiki/Republic_of_Crimea)) is specific, since Crimean legislature has declared conditional independence from Ukraine in 1992, but a referendum to confirm the decision was never held. Then in March 2014, the region organized an internationally disputed referendum where the majority voted in favor of independence of Crimea from Ukraine and joining Russia as a federal subject. Therefore, later during the same month newly unilaterally independent Republic of Crimea signed a treaty of accession to the Russian Federation. This is again an example of breakaway state since most of the international public nor Ukraine don’t formally recognize the changes of the borders, but it is also an example of unification of the Republic of Crimea with Russian Federation. 

We decided not to involve past or current from the [United Nations list of Non-Self-Governing Territories](https://en.wikipedia.org/wiki/United_Nations_list_of_Non-Self-Governing_Territories) in our analysis because of their special status and low number of citizens (while the lack of their codes in most of the world region and ethnic mappings was also a factor that disables their analysis).

## Regions with potential border changes

The regions that we will analyze in order to see if they show any similarities to the events that preceded previously observed types of border changes are the ones that already showed some internationally noticeable tendencies toward independence or unification. These are the regions of [Catalonia](https://en.wikipedia.org/wiki/Catalonia), [Donetsk](https://en.wikipedia.org/wiki/Donetsk), [Luhansk](https://en.wikipedia.org/wiki/Luhansk) and [Scotland](https://en.wikipedia.org/wiki/Scotland). Their circumstances will be explained in more detail in the later steps of the analysis.

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option('display.max_columns', 60)
pd.set_option('display.max_rows', 600)

#plotting params
%matplotlib inline
sns.set_style('darkgrid')
plt.rcParams['axes.labelsize'] = 14
plt.rcParams['axes.titlesize'] = 18
plt.rcParams['font.size'] = 12
plt.rcParams['figure.figsize'] = [16, 8]
plt.rcParams['axes.edgecolor'] = '0.15'
plt.rcParams['axes.linewidth']  = 1.25

In [2]:
#limit ourselves to a subset of data for initial event analysis
observed_years = [1990, 1991]

The information that we used to get the "creation date" of a country are the [*Gleditsch and Ward* lists of independent states and micronations](http://privatewww.essex.ac.uk/~ksg/statelist.html). We chose to use this dataset because it matches well with our previous textual analysis and *Gleditsch and Ward* country codes are included in the UCDP dataset.

We wrote a function *read_gw()* which reads our country information from the specified url with the correct encoding and names the columns appropriately.

In [3]:
#function for loading Gleditsch-Ward country codes from a specified url
def read_gw(url):
    return pd.read_csv(url, sep='\t', encoding = 'ISO-8859-1', header = None, 
                       names = ['gw_code', 'abbreviation', 'country_name', 'beginning', 'end'])

We wanted to download the list of independent states as well as the "tentative" list of micronations, some of which are important for our analysis such as Abkhazia, South Ossetia, Palau, Federated States of Micronesia, all of which declared independence after 1979.

In [4]:
gw_codes = read_gw('http://privatewww.essex.ac.uk/~ksg/data/iisystem.dat')
gw_micro = read_gw('http://privatewww.essex.ac.uk/~ksg/data/microstatessystem.dat')
country_codes = gw_codes.append(gw_micro, ignore_index = True)
country_codes.head(6)

gw_code abbreviation              country_name   beginning         end
0        2          USA  United States of America  01:01:1816  31:12:2012
1       20          CAN                    Canada  01:07:1867  31:12:2012
2       31          BHM                   Bahamas  10:07:1973  31:12:2012
3       40          CUB                      Cuba  20:05:1902  31:12:2012
4       41          HAI                     Haiti  01:01:1816  04:07:1915
5       41          HAI                     Haiti  15:08:1934  31:12:2012

We can see that the column *gw_code* can have same values for two entries (like Haiti above). Also, we have the *abbreviation* of the country (that, unfortunately, is not according to [ISO-3166 standard](https://www.iso.org/iso-3166-country-codes.html) which is the 3-letter country code that GDELT uses), the full name of the country and its "beginning" and "end".

We can also observe that this is a historical list of countries that existed from 1st of January 1816 until 31st of December 2012.

We also needed to convert the "beginning" and "end" of countries to an appropriate date format.

In [5]:
#convert columns to the right date format
country_codes['beginning'] = pd.to_datetime(country_codes['beginning'], format='%d:%m:%Y')
country_codes['end'] = pd.to_datetime(country_codes['end'], format='%d:%m:%Y')

As mentioned before, the 3-letter country abbreviations in the *Gleditsch and Ward* country list do not match the 3-letter ActorXCountryCode in the *GDELT* dataset, so we had to add another column, *alias*, where we put the codes of countries we need to observe.

Also, while looking at the dataset, we noticed that there were some inconcistencies within the *GDELT* dataset itself. For example, page 102, chapter 3.2.1 of the [CAMEO codes manual](http://data.gdeltproject.org/documentation/CAMEO.Manual.1.1b3.pdf) gives an example of how country codes for Serbia changed over time. It specifies that Serbia had the code YUGSRB before 1992, FRYSRB between 1992 and 2003 and SCGSRB from 2003 until 2006. However, while examining the dataset, we noticed that the code SCGSRB is used to refer to Serbia in 1979!

In [6]:
#dictionary of country codes and its aliases, will be completed and read from a file in the next milestone
country_alias = {
    #new countries
    'MLD': 'MDA',
    'KYR': 'KGZ',
    'LAT': 'LVA',
    'LIT': 'LTU',
    'GRG': 'GEO',
    'TAJ': 'TJK',
    'MAC': 'MKD|YUGMCD',
    'KZK': 'KAZ',
    'BOS': 'BIH|YUGBSN',
    'SLV': 'SVN|YUGSLN',
    'CRO': 'HRV|YUGCTA',
    'CZR': 'CZE',
    'SLO': 'SVK',
    'PAL': 'PLW',
    'ETM': 'TMP',
    'MNG': 'MTN',
    'SER': 'SRB',
    'KOS': 'SCGKOS|YUGKOS',
    'SOT': 'oss',
    'ABK': 'abk',
    #old countries
    'SAF': 'ZAF',
    'UKG': 'GBR',
    'SUD': 'SDN',
    'INS': 'IDN',
    'RUS': 'USR',
    'YUG': 'SCG',
    'CZE': 'CZA',
    'GRG': 'GEO'
}

#create an empty column
country_codes['alias'] = None
#fill the column values for the observed countries
for key in country_alias.keys():
    country_codes.loc[country_codes['abbreviation'] == key, 'alias'] = country_alias[key]

country_codes.tail()

gw_code abbreviation                    country_name  beginning  \
234      973          TUV                          Tuvalu 1978-10-01   
235      983          MSI                Marshall Islands 1986-10-21   
236      986          PAL                           Palau 1994-10-01   
237      987          FSM  Federated States of Micronesia 1986-11-03   
238      990          WSM             Samoa/Western Samoa 1962-01-01   

           end alias  
234 2012-12-31  None  
235 2012-12-31  None  
236 2012-12-31   PLW  
237 2012-12-31  None  
238 2012-12-31  None

Afterwards, for the initial analysis of a subset of data for Milestone 2, we restricted ourselves to the countries that were created after the 1st of January 1990 and sorted them by their formation date. 

In [7]:
#filter and sort countries
new_countries = country_codes[(country_codes['beginning'] > '1990-01-01')].sort_values(by = 'beginning')
new_countries

gw_code abbreviation                             country_name  beginning  \
140      565          NAM                                  Namibia 1990-03-21   
90       370          BLR                    Belarus (Byelorussia) 1991-08-25   
80       359          MLD                                  Moldova 1991-08-27   
179      703          KYR                          Kyrgyz Republic 1991-08-31   
180      704          UZB                               Uzbekistan 1991-08-31   
86       367          LAT                                   Latvia 1991-09-06   
88       368          LIT                                Lithuania 1991-09-06   
92       372          GRG                                  Georgia 1991-09-06   
84       366          EST                                  Estonia 1991-09-06   
178      702          TAJ                               Tajikistan 1991-09-09   
177      701          TKM                             Turkmenistan 1991-10-27   
69       343          MAC  Macedonia (Former Yugoslav Republic of) 1991-11-20   
89       369          UKR                                  Ukraine 1991-12-01   
181      705          KZK                               Kazakhstan 1991-12-16   
93       373          AZE                               Azerbaijan 1991-12-21   
91       371          ARM                                  Armenia 1991-12-21   
74       346          BOS                       Bosnia-Herzegovina 1992-04-27   
76       349          SLV                                 Slovenia 1992-04-27   
70       344          CRO                                  Croatia 1992-04-27   
57       316          CZR                           Czech Republic 1993-01-01   
58       317          SLO                                 Slovakia 1993-01-01   
131      531          ERI                                  Eritrea 1993-05-24   
236      986          PAL                                    Palau 1994-10-01   
210      860          ETM                               East Timor 2002-05-20   
68       341          MNG                               Montenegro 2006-06-03   
72       340          SER                                   Serbia 2006-06-05   
75       347          KOS                                   Kosovo 2008-02-17   
227      397          SOT                            South Ossetia 2008-08-26   
226      396          ABK                                 Abkhazia 2008-08-26   
157      626          SSD                              South Sudan 2011-07-09   

           end          alias  
140 2012-12-31           None  
90  2012-12-31           None  
80  2012-12-31            MDA  
179 2012-12-31            KGZ  
180 2012-12-31           None  
86  2012-12-31            LVA  
88  2012-12-31            LTU  
92  2012-12-31            GEO  
84  2012-12-31           None  
178 2012-12-31            TJK  
177 2012-12-31           None  
69  2012-12-31     MKD|YUGMCD  
89  2012-12-31           None  
181 2012-12-31            KAZ  
93  2012-12-31           None  
91  2012-12-31           None  
74  2012-12-31     BIH|YUGBSN  
76  2012-12-31     SVN|YUGSLN  
70  2012-12-31     HRV|YUGCTA  
57  2012-12-31            CZE  
58  2012-12-31            SVK  
131 2012-12-31           None  
236 2012-12-31            PLW  
210 2012-12-31            TMP  
68  2012-12-31            MTN  
72  2012-12-31            SRB  
75  2012-12-31  SCGKOS|YUGKOS  
227 2012-12-31            oss  
226 2012-12-31            abk  
157 2012-12-31           None

We also wanted to include information about the countries from which they were formed, and because of that we added columns *former_country*, *former_country_gw_code*, and *former_country_alias*.

We wrote a function *find_former()* that for each newly formed country finds its former country as well as its *gw_code* and *alias*.

In [8]:
former_dict = {
    'SAF': ['NAM'],
    'RUS': ['BLR', 'MLD', 'KYR', 'UZB', 'LAT', 'LIT', 'GRG', 'EST', 'TAJ', 'TKM', 'UKR', 'KZK', 'AZE', 'ARM'],
    'YUG': ['MAC', 'BOS', 'SLV', 'CRO', 'MNG', 'SER'],
    'CZE': ['CZR', 'SLO'],
    'ETH': ['ERI'],
    'USA': ['PAL'],
    'INS': ['ETM'],
    'SER': ['KOS'],
    'GRG': ['SOT', 'ABK'],
    'SUD': ['SSD']
}

def find_former(row):
    for former in former_dict:
        if row['abbreviation'] in former_dict[former]:
            return pd.Series([former,
                              country_codes.loc[country_codes['abbreviation'] == former]['gw_code'].iloc[0],
                              country_codes.loc[country_codes['abbreviation'] == former]['alias'].iloc[0]])
    return pd.Series([None, None, None])

#fill the new columns with the returned Series values
new_countries[['former_country', 'former_country_gw_code', 'former_country_alias']] = new_countries.apply(lambda row: find_former(row), axis = 1)
new_countries.head()

gw_code abbreviation           country_name  beginning        end alias  \
140      565          NAM                Namibia 1990-03-21 2012-12-31  None   
90       370          BLR  Belarus (Byelorussia) 1991-08-25 2012-12-31  None   
80       359          MLD                Moldova 1991-08-27 2012-12-31   MDA   
179      703          KYR        Kyrgyz Republic 1991-08-31 2012-12-31   KGZ   
180      704          UZB             Uzbekistan 1991-08-31 2012-12-31  None   

    former_country  former_country_gw_code former_country_alias  
140            SAF                     560                  ZAF  
90             RUS                     365                  USR  
80             RUS                     365                  USR  
179            RUS                     365                  USR  
180            RUS                     365                  USR

After creating the initial DataFrame of observed countries, we are ready to load our sample GDELT dataset and UCDP datasets.

We chose not to use the cluster for accessing the GDELT dataset at this stage as it the version of the dataset which contains a **greatly reduced** number of columns compared to the regular GDELT dataset. Instead we chose to download only a couple of years of GDELT events and work with a table that has the full number of columns as that data will actually be used until the end of the project. Working with a database we would later have no use of made no sense to us and therefore we chose this approach.

The header for the GDELT tables can be found in a [webpage](https://www.gdeltproject.org/data/lookups/CSV.header.historical.txt) separate from the table contents themselves. We downloaded that data and stored it as our *column_names* variable.

In [9]:
header_file = open('data/GDELT/CSV.header.historical.txt', 'r')
column_names = header_file.read().strip('\n').split('\t')

FileNotFoundError: [Errno 2] No such file or directory: 'data/GDELT/CSV.header.historical.txt'

When we got the headers, we were ready to load the data itself. For now, we limited ourselves to the years of 1990 and 1991, during which the creation of new countries was the most frequent.

In [ ]:
events90 = pd.read_csv('data/GDELT/1990.csv', sep='\t', header = None, names = column_names)
events91 = pd.read_csv('data/GDELT/1991.csv', sep='\t', header = None, names = column_names)
events = events90.append(events91)
events.shape

We can see that we have around 2.6 million entries in total for the years 1990 and 1991.

After loading, we had to convert the dates in the columns *SQLDATE*, *MonthYear* and *Year* in order to process them easier.

In [ ]:
#convert columns to the right date format
events['SQLDATE'] = pd.to_datetime(events['SQLDATE'], format='%Y%m%d')
events['MonthYear'] = pd.to_datetime(events['MonthYear'], format='%Y%m')
events['Year'] = pd.to_datetime(events['Year'], format='%Y')

Now we are ready to begin the analysis of our data sample. We created a search criteria constructed from all the country abbreviations and aliases that will be checked as a regex in the *str.contains()* function.

We filtered the events by checking whether the columns *Actor1CountryCode*, *Actor2CountryCode*, *Actor1EthnicCode* or *Actor2EthnicCode* contain our search criteria.

For the next milestone, we plan on including the columns *Actor1Geo_CountryCode*, *Actor2Geo_CountryCode* and *ActionGeo_CountryCode* to our analysis. These columns also contain country codes but these codes are in a [completely different format](https://en.wikipedia.org/wiki/FIPS_10-4) from all the previous ones...

In [ ]:
search_criteria = '|'.join(new_countries['abbreviation']) + '|'.join(new_countries['alias'].dropna())
#filter by search criteria
observed_events = events.loc[(events['Actor1CountryCode'].str.contains(search_criteria, na=False))|
                             (events['Actor2CountryCode'].str.contains(search_criteria, na=False))|
                             (events['Actor1EthnicCode'].str.contains(search_criteria, na=False))|
                             (events['Actor2EthnicCode'].str.contains(search_criteria, na=False))]
observed_events.shape

We can observe that the number of events has reduced significantly, to around 186000, which is a more managable number.

Next, we can show the percentage of values in the columns that are not NaN.

In [ ]:
#get percentage of non-NaN values
print(observed_events['Actor1CountryCode'].count()/len(observed_events) * 100)
print(observed_events['Actor2CountryCode'].count()/len(observed_events) * 100)
print(observed_events['Actor1EthnicCode'].count()/len(observed_events) * 100)
print(observed_events['Actor2EthnicCode'].count()/len(observed_events) * 100)
print(observed_events['Actor1Geo_CountryCode'].count()/len(observed_events) * 100)
print(observed_events['Actor2Geo_CountryCode'].count()/len(observed_events) * 100)
print(observed_events['ActionGeo_CountryCode'].count()/len(observed_events) * 100)

We see that while the country codes are mostly filled everywhere, ethnic codes exist in less then 1% of our observed events.

Now we can see the distribution of the different number of root events, grouped by countries. The observed column is the *EventRootCode*, which assigns a general classification to an event. The columns *EventBaseCode* and *EventCode* further distinguish between the events.

In [ ]:
event_counts = observed_events.groupby(by = ['Actor1CountryCode', 'EventRootCode']).size()

Here we showed event statistics for three countries: Armenia, Moldova and USSR/Russia.

In [ ]:
arm_event_count = pd.DataFrame(event_counts.loc['ARM'], columns = ['Event count'])
mda_event_count = pd.DataFrame(event_counts.loc['MDA'], columns = ['Event count'])
rus_event_count = pd.DataFrame(event_counts.loc['RUS'], columns = ['Event count'])

We created a simple function that plots our the counts of different types of events for the given DataFrame and country name.

In [ ]:
def plot_event_count(df, country_name):
    plt.figure()
    ax = sns.barplot(x = df.index, y = 'Event count', data = df)
    #set the plot title
    ax.set_title(country_name + ' event count')

In [ ]:
plot_event_count(arm_event_count, 'Armenia')
plot_event_count(mda_event_count, 'Moldova')
plot_event_count(rus_event_count, 'USSR/Russia')

The most striking insight from these plots is that the distribution of event types 4(*Consult*) and 5(*Engage in diplomatic cooperation*) are similarly distributed. We can also see that Armenia has proportionally a much greater number of events marked as 19(*Fight*) compared to Moldova and USSR/Russia. Besides that, we can see that the total number of events is much less for Moldova than for USSR/Russia.  


In these plots, events are designated by their code. In the later stages of the implementation, we plan on making a categoric variable from the event codes using [this dataset](https://www.gdeltproject.org/data/lookups/CAMEO.eventcodes.txt).

Now we can try adding the grouping of events by dates to show their temporal distribution.

In [ ]:
event_time_dist = events.groupby(by = ['Actor1CountryCode', 'MonthYear', 'EventRootCode']).size()

Here we showed event statistics for the same three countries: Armenia, Moldova and USSR/Russia. As an example, we chose to plot events with root codes 10(*Demand*) and 14(*Protest*).

In [ ]:
#transform the DataFrame with unstack() for easier plotting
arm_event_time = pd.DataFrame(event_time_dist.loc['ARM'], columns = ['Event count']).loc(axis=0)[:, [10, 14]].unstack(1, fill_value = 0)
rus_event_time = pd.DataFrame(event_time_dist.loc['RUS'], columns = ['Event count']).loc(axis=0)[:, [10, 14]].unstack(1, fill_value = 0)
mda_event_time = pd.DataFrame(event_time_dist.loc['MDA'], columns = ['Event count']).loc(axis=0)[:, [10, 14]].unstack(1, fill_value = 0)

We created a function *plot_event_time()* that plots our temporal data as a line plot.

In [ ]:
def plot_event_time(df, country_name):
    ax = df.plot()
    ax.set_title(country_name + ' events distribution')

In [ ]:
#sns.tsplot(time = rus_event_time.index, data = rus_event_time)
plot_event_time(arm_event_time, 'Armenia')
plot_event_time(mda_event_time, 'Moldova')
plot_event_time(rus_event_time, 'USSR/Russia')

We see that the number of 10-*Demand* event types increases for all three countries around August and September of 1991, which coincides with some of the [key events](https://en.wikipedia.org/wiki/1991_Soviet_coup_d%27%C3%A9tat_attempt) that led to the final dissolution of the USSR. Astoundingly, we do not see a large variation in the number of 14-*Protest* events during this timeframe within any of the three countries.

Above shown is just a beginning of analysis that we would initally do.

A way to enrich our analysis is to use *UCDP Conflict Termination Dataset* and *UCDP External Support Dataset*. With these datasets, we can evaluate how the number of, for example, *Fight* type events correlates with the observed termination of conflict inside a country.

In [ ]:
conflict_termination = pd.read_excel('data/UCDP/ucdp-term-conf-2015.xlsx')

After loading the *UCDP Conflict Termination Dataset*, we filtered it to fit our currently observing years of 1990 and 1991 and for the matching parties to correspond to the *Gleditsch and Ward* country numbers of our observed countries.

In [ ]:
conflict_term_90_91 = conflict_termination.loc[conflict_termination['Year'].isin(observed_years) & 
                                              (conflict_termination['GWNoA'].isin(new_countries['gw_code'].tolist())|
                                               conflict_termination['GWNoA2nd'].isin(new_countries['gw_code'].tolist())|
                                               conflict_termination['GWNoB'].isin(new_countries['gw_code'].tolist())|
                                               conflict_termination['GWNoB2nd'].isin(new_countries['gw_code'].tolist()))]                                                  
conflict_term_90_91

We can see that only two conflicts ended in 1991, the first one being between the *Government of Georgia* and the *National Guard and Mkhedrioni*, and the second one being between the *Government of Azerbaijan* on one side and the *Republic of Nagorno-Karabakh* with the support of *Government of Armenia* on the other side.

We extracted the IDs of these conflicts in an array so that we could filter the *UCDP External Support Dataset* by those conflict IDs. The conflict IDs in the two *UCDP* datasets slightly differ.

However, the conflict IDs in the *Conflict Termination Dataset* start with "1-" so we removed it in order to make it compatible with the *External Support Dataset*. 

In [ ]:
observed_conflict_ids = conflict_term_90_91['ConflictId'].apply(lambda x: x.split('-')[1]).tolist()
observed_conflict_ids

In [ ]:
external_support = pd.read_excel('data/UCDP/extsup_small.xls')
external_support.head()

After loading the *UCDP External Support Dataset*, we filtered it by our observed years of 1990 and 1991 and by the conflict IDs we obtained in previous steps.

In [ ]:
observed_support = external_support.loc[external_support['ywp_year'].isin(observed_years) &
                                        external_support['conflictID'].isin(observed_conflict_ids)]
observed_support

This DataFrame shows the external support for all the conflicted parties involved in a conflict. In the *Government of Azerbaijan - Republic of Nagorno-Karabakh* conflict only the the Republic of Nagorno-Karabakh received support, while the Government of Azerbaijan did not.
We can characterize the *Government of Georgia* - *National Guard and Mkhedrioni* conflict as being completely internal, as no conflicted party received known or alleged external support of any kind. 

We can observe that in the above DataFrame existed only one provider of external support, and that was Armenia, which supported the Republic of Nagorno-Karabakh.

We can see how it supported the Republic of Nagorno-Karabakh by printing the *external_type* and *external_comment* fields:

In [ ]:
print(observed_support['external_type'].dropna().tolist()[0])
print(observed_support['external_comment'].dropna().tolist()[0])

This finishes our sample analysis for this milestone, which hopefully demonstrates our ability to manipulate the different kinds of data we obtained from various sources that are relevant to our later analysis. The planned future steps of our analysis are outlined in the README file of the project.

<a id='m3'></a>
# Milestone 3

In [ ]:
#visualization
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import plotly
import plotly.plotly as py
from plotly.graph_objs import *
import IPython

#data_manipulation
import pandas as pd
import numpy as np
import networkx as nx
from networkx.algorithms import bipartite

#ML
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn import tree

from math import pi

plotly.tools.set_credentials_file(username='MilenaFilipovic', api_key='AiBlKAOtJ0Admy3u8e2u')
plt.rcParams['figure.figsize'] = [16, 8]
pd.options.display.max_rows = 999

## Border changes

As there is no comprehensive list of all border changes that occured during the last 40 years, we decided to create one ourselves. We believe that the list we compiled is pretty extensive and, if maybe not entirely complete in scope, it is enough for our analysis.

We created our list in a spreadsheet and exported it to csv, which we then loaded into the notebook.

In [ ]:
observed_countries = pd.read_csv('observed_countries.csv')
observed_countries

<a id='first_df'></a>
In the previous frame, we can see that we have a list of 54 countries that were created during the last 40 years. The dataframe is organised in the following manner:
* **Country**: The name of the newly created country
* **Type**: Type of border change
* **Year**: The name of the newly created country
* **Country before border change**: The name of the "parent" country
* **CAMEO new**: The CAMEO codes of the newly created country, with aliases for the same country; some entries also contain primary ethnic group in the country
* **CAMEO old**: The CAMEO codes of the parent country
* **FIPS new**: The FIPS codes of the newly created country; some entries also contain regions of a parent country they were a part of
* **FIPS old**: The FIPS codes of the parent country
* **Description**: A short description of border change

When we completed writing this list, we were ready to extract some events from the *GDELT dataset*. Sadly, even after all our pleas, the full *GDELT v1* dataset was not available on the cluster. Initially, *GDELT v1* was uploaded with a reduced number of columns, while the *GDELT v2* was uploaded with events starting from 2015, which was insufficient for our analysis.

Fortunately, *GDELT* was available on the [Google BigQuery](https://cloud.google.com/bigquery/) platform, and we could write our own SQL queries which would extract the data we needed.

To obtain this data, we used the information contained in the CAMEO and FIPS codes of each country. We created a function *print_country_codes()* which extracts and combines the codes of the same type for all countries.

In [ ]:
def print_country_codes(country_df):
    #get a list of all FIPS codes
    fips_country_codes = '|'.join(set(('|'.join(country_df['FIPS new'].dropna()) + '|' + '|'.join(country_df['FIPS old'].dropna())).split('|')))
    #get a list of all CAMEO codes
    cameo_codes = set(('|'.join(country_df['CAMEO new'].dropna()) + '|' + '|'.join(country_df['CAMEO old'].dropna())).split('|'))
    
    #divide CAMEO codes to country codes and ethnic codes
    cameo_country_codes = [code for code in cameo_codes if not code.islower()]
    cameo_ethnic_codes = '|'.join(cameo_codes - set(cameo_country_codes))
    cameo_country_codes = '|'.join(cameo_country_codes)
    
    #sort code lists
    fcc = fips_country_codes.split('|')
    fcc.sort()        
    ccc = cameo_country_codes.split('|')
    ccc.sort()    
    cec = cameo_ethnic_codes.split('|')
    cec.sort()
    
    #print them
    print('FIPS country codes:')
    print('|'.join(fcc)+'\n')
    
    print('CAMEO country codes:')
    print('|'.join(ccc)+'\n')
    
    print('CAMEO ethnic codes:')
    print('|'.join(cec)+'\n')

In [ ]:
print_country_codes(observed_countries)

When this was done, we were ready to execute our queries on the BigQuery platform.

Firstly, we chose to extract events for all countries. For each country participating in a specific border change, we extracted events relating to the participating countries since the beginning of the dataset(1979) until 5 years after the border change. The SQL script for this looked something like this:

```SQL
SELECT * FROM [gdelt-bq:full.events] 
WHERE
(YEAR<1998 AND
(REGEXP_MATCH (CONCAT(IFNULL(Actor1Code, '+'), '_', IFNULL(Actor1CountryCode, '+'), '_', IFNULL(Actor2Code, '+'), '_', IFNULL(Actor2CountryCode, '+')), r'GEO') OR
REGEXP_MATCH (CONCAT(IFNULL(Actor1Geo_CountryCode, '+'), '_', IFNULL(Actor1Geo_ADM1Code, '+'), '_', IFNULL(Actor2Geo_CountryCode, '+'), '_', IFNULL(Actor2Geo_ADM1Code, '+')), r'GG02|GG') OR REGEXP_MATCH (CONCAT(IFNULL(Actor1EthnicCode, '+'), '_', IFNULL(Actor2EthnicCode, '+')), r'abk')))
OR
(YEAR<1986 AND
(REGEXP_MATCH (CONCAT(IFNULL(Actor1Code, '+'), '_', IFNULL(Actor1CountryCode, '+'), '_', IFNULL(Actor2Code, '+'), '_', IFNULL(Actor2CountryCode, '+')), r'ATG|GBR') OR
REGEXP_MATCH (CONCAT(IFNULL(Actor1Geo_CountryCode, '+'), '_', IFNULL(Actor1Geo_ADM1Code, '+'), '_', IFNULL(Actor2Geo_CountryCode, '+'), '_', IFNULL(Actor2Geo_ADM1Code, '+')), r'AC|UK')))
OR...
```

We stored those events in a separate table and then decided to group the events by their ActorCountryCode and the EventRootCode:

```SQL
SELECT
  Actor1CountryCode AS ActorCountryCode,
  EventRootCode,
  COUNT(EventRootCode) AS EventCount,
  AVG(GoldsteinScale) AS AvgGoldScore,
  AVG(AvgTone) AS AvgTone
FROM
  [GDELT.gdelt_filtered]
WHERE
  REGEXP_MATCH (Actor1CountryCode, r'ARM|ATG|AZE|BIH|BIHBHF|BIHSRP|BLR|BLZ|BRN|CRO|CYP|CZA|CZE|DEU|ERI|EST|ETH|FRA|FRY|FRYKSV|FRYMTN|FRYSRB|FSM|GBR|GEO|GMB|GME|GMW|HRV|IDN|IRQ|ISR|KAZ|KGZ|KHM|KIR|KNA|KWT|LCA|LTU|LVA|MDA|MHL|MKD|MTN|NAM|PAL|PLW|PSE|RUS|RUSCNY|SCG|SCGKSV|SCGMTN|SCGSRB|SDN|SEN|SER|SOM|SRB|SRBKSV|SSD|SVK|SVN|TJK|TKM|TMP|UKR|USA|USR|UZB|VCT|VNM|VUT|YEM|YMN|YMS|YUG|YUGBSN|YUGCTA|YUGKSV|YUGMCD|YUGMTN|YUGSLN|YUGSRB|ZAF|ZWE')
  GROUP BY
  Actor1CountryCode,
  EventRootCode
ORDER BY
  Actor1CountryCode,
  EventRootCode
```

When we got the results of the query, we downloaded them as as .csv file. We did this for both actors (Actor1CountryCode, Actor2CountryCode), as well as both types of codes (Actor1Geo_CountryCode, Actor2Geo_CountryCode), by skipping the events we already extracted with the previous query:

```SQL
SELECT
  Actor1Geo_CountryCode AS ActorGeo_CountryCode,
  EventRootCode,
  COUNT(EventRootCode) AS EventCount,
  AVG(GoldsteinScale) AS AvgGoldScore,
  AVG(AvgTone) AS AvgTone
FROM
  [GDELT.gdelt_filtered]
WHERE
  REGEXP_MATCH (Actor1Geo_CountryCode, r'A56|AC|AJ|AJ03|AJ14|AJ18|AJ26|AJ28|AJ43|AJ55|AJ57|AJ61|AJ64|AJ65|AJ69|AM|BH|BK|BO|BX|BZ|CB|CY|CY01|CY02|CZ|CZ01|CZ02|EN|ER|ET|ET04|EZ|FM|FR|GA|GC|GE|GG|GG02|GG05|GG24|GM|GZ|HR|ID|ID27|IS|IZ|KG|KR|KU|KV|KZ|LG|LH|LO|MD|MD04|MD09|MD19|MD25|MD33|MD36|MD43|MJ|MK|MW|NH|NQ20|NQ25|NQ30|OD|PS|RB|RI|RM|RS|RS12|SC|SF|SG|SI|SO|SO12|SO15|SO21|SO22|SR|ST|SU|TI|TT|TX|UK|UP|UP11|UR|UR01|UR02|UR03|UR04|UR05|UR06|UR07|UR08|UR09|UR10|UR11|UR12|UR13|UR14|UR15|US|UZ|VC|VM|WA|WE|YE|YI|YI01|YI02|YM|YO|YO01|YO02|YO03|YO05|YO06|YO07|YS|ZI')
  AND NOT REGEXP_MATCH (Actor1CountryCode, r'ARM|ATG|AZE|BIH|BIHBHF|BIHSRP|BLR|BLZ|BRN|CRO|CYP|CZA|CZE|DEU|ERI|EST|ETH|FRA|FRY|FRYKSV|FRYMTN|FRYSRB|FSM|GBR|GEO|GMB|GME|GMW|HRV|IDN|IRQ|ISR|KAZ|KGZ|KHM|KIR|KNA|KWT|LCA|LTU|LVA|MDA|MHL|MKD|MTN|NAM|PAL|PLW|PSE|RUS|RUSCNY|SCG|SCGKSV|SCGMTN|SCGSRB|SDN|SEN|SER|SOM|SRB|SRBKSV|SSD|SVK|SVN|TJK|TKM|TMP|UKR|USA|USR|UZB|VCT|VNM|VUT|YEM|YMN|YMS|YUG|YUGBSN|YUGCTA|YUGKSV|YUGMCD|YUGMTN|YUGSLN|YUGSRB|ZAF|ZWE')
GROUP BY
  Actor1Geo_CountryCode,
  EventRootCode
ORDER BY
  Actor1Geo_CountryCode,
  EventRootCode
```

Combining the events that contain a CAMEO country code and a FIPS country code relating to a specific country proved useful, as we obtained a decent number of different events with the FIPS filtering.
We tried to do the same with other columns that may contain relevant information (ActorCode, ActorEthnicCode, ActorGeo_ADM1Code), but this yielded no more events.
 
The next step was to load and combine the datasets we obtained from BigQuery into our notebook. For that, we created functions *load_both_actor_datasets()* and *merge_actor_columns()*.

We also created a mapping for the EventRootCode column to its corresponding textual meaning.

In [ ]:
event_codes = {
1: 'Make Public Statement',
2: 'Appeal',
3: 'Express Intent To Cooperate',
4: 'Consult',
5: 'Engage In Diplomatic Cooperation',
6: 'Engage In Material Cooperation',
7: 'Provide Aid',
8: 'Yield',
9: 'Investigate',
10: 'Demand',
11: 'Disapprove',
12: 'Reject',
13: 'Threaten',
14: 'Protest',
15: 'Exhibit Military Posture',
16: 'Reduce Relations',
17: 'Coerce',
18: 'Assault',
19: 'Fight',
20: 'Engage In Unconventional Mass Violence'}

In [ ]:
def load_both_actor_datasets(actor1_file, actor2_file, index_col):
    #load both datasets and drop nulls
    actor1_grouping = pd.read_csv(actor1_file)
    actor2_grouping = pd.read_csv(actor2_file)
    actor1_grouping.dropna(inplace = True)
    actor2_grouping.dropna(inplace = True)
    #change event root codes to their textual format
    actor1_grouping['EventRootCode'] = actor1_grouping['EventRootCode'].apply(lambda x: event_codes.get(x))
    actor2_grouping['EventRootCode'] = actor2_grouping['EventRootCode'].apply(lambda x: event_codes.get(x))
    #set index
    actor1_grouping.set_index(keys = index_col, inplace = True)
    actor2_grouping.set_index(keys = index_col, inplace = True)
    #merge tables
    merged_actors = pd.merge(actor1_grouping, actor2_grouping, left_index=True, right_index=True, how='outer')
    return merged_actors

def merge_actor_columns(merged_df):
    merged_df['EventCount'] = merged_df['EventCount_x'].fillna(0).astype('int64') + merged_df['EventCount_y'].fillna(0).astype('int64')
    merged_df['AvgGoldScore'] = (merged_df['EventCount_x'].fillna(0).astype('int64') * merged_df['AvgGoldScore_x'].fillna(0)
                                   +merged_df['EventCount_y'].fillna(0).astype('int64') * merged_df['AvgGoldScore_y'].fillna(0))/merged_df['EventCount']
    merged_df['AvgTone'] = (merged_df['EventCount_x'].fillna(0).astype('int64') * merged_df['AvgTone_x'].fillna(0)
                                   +merged_df['EventCount_y'].fillna(0).astype('int64') * merged_df['AvgTone_y'].fillna(0))/merged_df['EventCount']
    merged_df.drop(['EventCount_x', 'AvgGoldScore_x', 'AvgTone_x', 'EventCount_y', 'AvgGoldScore_y', 'AvgTone_y'], axis=1, inplace = True)

Firstly, we loaded the datasets for both actors with events extracted from CAMEO country codes.

In [ ]:
cameo_grouped = load_both_actor_datasets(actor1_file = 'Actor1CountryCode_grouping.csv',
                                         actor2_file = 'Actor2CountryCode_grouping.csv',
                                         index_col = ['ActorCountryCode', 'EventRootCode'])
cameo_grouped.head()

Here, we see that we have the number of events for each country and a specific type of event, divided by actors.

We also have the columns **AvgGoldScore** and **AvgTone**. The column **AvgGoldScore** reprsents the average value of  the potential impact that this type of event will have on the stability of a country. The column **AvgTone** represents the average “tone” of all documents containing one or more mentions of this type of event.

A more detailed description of the Goldstain Scale and the Average Tone of an event can be found in the [GDELT Data Format Codebook](http://data.gdeltproject.org/documentation/GDELT-Data_Format_Codebook.pdf).

Next, we wanted to merge the columns for both actors. We added the number of events for each actor and calculated the **AvgGoldScore** and **AvgTone** appropriately.

In [ ]:
merge_actor_columns(cameo_grouped)
cameo_grouped.head()

When this was done, we did the same for the events that we extracted for the FIPS country code.

In [ ]:
fips_grouped = load_both_actor_datasets(actor1_file = 'Actor1FIPSCountryCode_grouping.csv',
                                        actor2_file = 'Actor2FIPSCountryCode_grouping.csv',
                                        index_col = ['ActorGeo_CountryCode', 'EventRootCode'])
merge_actor_columns(fips_grouped)
fips_grouped.head()

We can see that the number of the events we extraced with FIPS codes is not huge, but we still felt that it would provide more accurate results then just analyzing the events extracted only by the CAMEO codes of the observed countries.

Now, we were ready to unify the events we obtained by extracting events by CAMEO codes, as well as those we extracted by FIPS codes. To get this, we needed to create a matching between CAMEO and FIPS codes of a specific country. Luckily, we had that in mind and we created our initial dataframe in such a way that we can obtain a correct matching by observing the first FIPS or CAMEO code related to that border change, for example:
* Armenia CAMEO: ARM|arm, Armenia FIPS: AM|UR01; created matching: AM->ARM.

In [ ]:
def get_main_cameo(row, observed_column):
    #if a code in the observed row exists
    if pd.notnull(row[observed_column]):
        #get the first code
        return row[observed_column].split('|')[0]
    else:
        return None

In [ ]:
#create new columns with the main CAMEO and FIPS codes of new and parent countries
observed_countries['Main CAMEO new'] = observed_countries.apply(lambda row: get_main_cameo(row, observed_column = 'CAMEO new'),axis=1)
observed_countries['Main CAMEO old'] = observed_countries.apply(lambda row: get_main_cameo(row, observed_column = 'CAMEO old'),axis=1)
observed_countries['Main FIPS new'] = observed_countries.apply(lambda row: get_main_cameo(row, observed_column = 'FIPS new'),axis=1)
observed_countries['Main FIPS old'] = observed_countries.apply(lambda row: get_main_cameo(row, observed_column = 'FIPS old'),axis=1)

When we obtained the main CAMEO and FIPS codes we had created a dataframe contained only from those columns.

In [ ]:
#create a list of code pairings, firstly creating a list of new codes,
#and then append parent country pairings, by dropping duplicates,
#as same pairs can occur in this list (more countries have a same parent)
main_codes = observed_countries[['Main CAMEO new', 'Main FIPS new']]\
             .rename(columns = {'Main CAMEO new': 'Main CAMEO', 'Main FIPS new': 'Main FIPS'})\
             .append(observed_countries[['Main CAMEO old', 'Main FIPS old']].drop_duplicates()\
                     .rename(columns = {'Main CAMEO old': 'Main CAMEO', 'Main FIPS old': 'Main FIPS'}), ignore_index = True)
main_codes.dropna(inplace = True)

We still had some issues that we had to solve manually, as Serbia is both a new country (created from Serbia and Montenegro), and a parent (of Kosovo). Also, Palestine is refered with fips codes *GZ* and *WE*.

In [ ]:
main_codes = pd.concat([main_codes, pd.DataFrame({'Main CAMEO': ['SRB', 'PSE'], 'Main FIPS': ['RB', 'WE']}, index = [80, 81])])

Now we can change the FIPS event code to its corresponding CAMEO one with the function *change_fips_to_cameo()*.

In [ ]:
def change_fips_to_cameo(row, code_mapping):
    #if both codes are present in the current row, replace its FIPS country code to the CAMEO country code
    if (row['ActorGeo_CountryCode'] is s for s in code_mapping['Main FIPS'].unique()):
        if(len(code_mapping.loc[code_mapping['Main FIPS'] == row['ActorGeo_CountryCode']]['Main CAMEO'].values) > 0):
            row['ActorGeo_CountryCode'] = code_mapping.loc[code_mapping['Main FIPS'] == row['ActorGeo_CountryCode']]['Main CAMEO'].values[0]
    return row

#we can't modify the index directly so we have to drop it
fips_grouped.reset_index(inplace = True)
#apply the function
fips_grouped = fips_grouped.apply(lambda row: change_fips_to_cameo(row, main_codes), axis = 1)
#rename the code column so that it corresponds to the CAMEO column name in the other DataFrame
fips_grouped.rename(columns={'ActorGeo_CountryCode': 'ActorCountryCode'}, inplace = True)
#return the index
fips_grouped.set_index(['ActorCountryCode', 'EventRootCode'], inplace=True)
#show the code index values
fips_grouped.index.levels[0]

We can see that we have successfully transformed all of the FIPS codes to their CAMEO counterparts.

Finally, having obtained the same codes across both DataFrames (and therefore the same indexes), we can merge the two in a single DataFrame.

merged_countries = pd.merge(cameo_grouped, fips_grouped, left_index=True, right_index=True, how='outer')
merge_actor_columns(merged_countries)
merged_countries.isnull().values.any()

Having no null values means that we have completely and successfully created a single DataFrame, on which we can perform further analysis.

In [ ]:
#save the copy of a DataFrame for later
merged_countries_original = merged_countries.copy()